lib/compilation_engine.rb

```diff
@@ -6,18 +6,29 @@ def initialize(input, output)
 
   def compile_class
     @output.puts("<class>")
-    @tokenizer.has_more_tokens?
-    @tokenizer.advance
+    @tokenizer.has_more_tokens? && @tokenizer.advance
     @output.puts("  <keyword> class </keyword>")
 
-    @tokenizer.has_more_tokens?
-    @tokenizer.advance
+    @tokenizer.has_more_tokens? && @tokenizer.advance
     @output.puts("  <identifier> #{@tokenizer.identifier} </identifier>")
 
-    @output.puts <<~HEREDOC
-        <symbol> { </symbol>
-        <symbol> } </symbol>
-      </class>
-    HEREDOC
+    @tokenizer.has_more_tokens? && @tokenizer.advance
+    @output.puts("  <symbol> { </symbol>")
+
+    @tokenizer.has_more_tokens? && @tokenizer.advance
+    if @tokenizer.token_type == :KEYWORD
```

Comment:
Can you give me an example of a grammar that requires no lookahead, multiple lookahead, and backtracking?

Reply:
Sure!

Here’s a grammar (borrowing the notation from EoCS) which requires no lookahead:

> phoneNumber: countryCode areaCode exchangeCode lineNumber
> countryCode: `'+'` digit digit digit
> areaCode: `'('` digit digit digit `')'`
> exchangeCode: digit digit digit
> lineNumber: digit digit digit digit

Hopefully you can imagine implementing methods to parse _phoneNumber_, _countryCode_ etc from a stream of tokens (so the tokenizer has already skipped whitespace and, in this case, turned each digit into a separate token). No lookahead implies there’s never any decision to make about what to do next — each method would just be a sequence of calls to `#advance` and/or other parsing methods. It’s not very interesting but it is sometimes what we do when we want to parse a fixed format.

Here’s a tiny example of a grammar which requires two-token lookahead:

> program: block | emptyBlock
> block: `'{'` `'hello'` `'}'`
> emptyBlock: `'{'` `'}'`

The language described by this grammar actually only has two possible programs — `{ }` and `{ hello }` — but when you’re parsing the _program_ production you can’t decide whether to parse _block_ or _emptyBlock_ just by looking at the next token, because it’s '`{'` in both cases, so you need to check the token after that to decide what to do.

(You might already have noticed that we could actually rewrite this grammar in a way which avoids the need to look ahead more than one token, but that’s not always possible.)

Here’s a tiny example of a grammar which causes the parser to backtrack:

> statement: greeting recipient
> greeting: `'hello'` `'world'` | `'hello'`
> recipient: `'world'`

Imagine we’re parsing the string `'hello world'`. The method which parses the _greeting_ production can’t make an immediate decision about whether to read `'hello world'` or just `'hello'`, even by looking ahead at the next token, so it will have to pick one and continue. If it picks `'hello world'` then the parsing will fail later when we try to parse the _recipient_ production — it needs to read the `'world'` token but that’s already been dealt with. So the parser needs to abort, rewind the token stream, and try the other choice. This is complicated to implement correctly.

In all cases it’s ambiguity that causes the problem. Fortunately Jack’s grammar has been designed to be completely unambiguous so we don’t have to deal with these issues.